overlord/snapshotstate/backend/backend.go

```diff
@@ -0,0 +1,220 @@
+package backend
+
+import (
+	"archive/zip"
+	"crypto"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const (
+	archiveName  = "archive.tgz"
+	metadataName = "meta.json"
+	metaHashName = "meta.sha3_384"
+
+	userArchivePrefix = "user/"
+	userArchiveSuffix = ".tgz"
+)
+
+// Iter loops over all valid snapshots in the snapshots directory,
+// applying the given function to each. It is that function's
+// responsibility to close the snapshot it's passed. If the function
+// returns error, iteration is stopped (and if the error isn't EOF,
+// it's returned as the error of the iterator).
+func Iter(ctx context.Context, f func(*Reader) error) error {
+	if err := ctx.Err(); err != nil {
+		return err
+	}
+
+	dir, err := os.Open(dirs.SnapshotDir)
+	if osutil.IsDirNotExist(err) {
+		// no dir -> no snapshots
+		return nil
+	}
+
+	var names []string
+	for err == nil {
+		names, err = dir.Readdirnames(100)
+		for _, name := range names {
+			if err = ctx.Err(); err != nil {
+				break
+			}
+
+			var rsh *Reader
+			// wish dirs had Open() that did openat
+			rsh, err = Open(filepath.Join(dirs.SnapshotDir, name))
+			if err != nil {
+				err = nil
+				continue
+			}
+			err = f(rsh)
+			if err != nil {
+				break
+			}
+			err = rsh.Close()
+			if err != nil {
+				break
+			}
+		}
+	}
+
+	if err == io.EOF {
+		err = nil
+	}
+
+	return err
+}
+
+// List valid snapshots
+//
+// The snapshots are closed before returning.
+func List(ctx context.Context, snapshotID uint64, snapNames []string) ([]client.SnapshotGroup, error) {
+	shotsmap := map[uint64][]*client.Snapshot{}
+	err := Iter(ctx, func(sh *Reader) error {
+		if snapshotID == 0 || sh.ID == snapshotID {
+			if len(snapNames) == 0 || strutil.ListContains(snapNames, sh.Snap) {
+				shotsmap[sh.ID] = append(shotsmap[sh.ID], &sh.Snapshot)
+			}
+		}
+		return nil
+	})
+
+	groups := make([]client.SnapshotGroup, 0, len(shotsmap))
+	for id, shots := range shotsmap {
+		sort.Sort(bySnap(shots))
+		groups = append(groups, client.SnapshotGroup{ID: id, Snapshots: shots})
+	}
+
+	sort.Sort(byID(groups))
+
+	return groups, err
+}
+
+var isValidVersion = regexp.MustCompile("^[a-zA-Z0-9.+~-]{1,32}$").MatchString
```

Comment:
What are the `~` and `+` characters for?

Reply:
this regexp is from snapcraft (and I mentioned that it was weird that they let a thing have a version of `README.~1~`... :-) )